Firestore/core/src/firebase/firestore/model/mutation.h

```diff
@@ -0,0 +1,148 @@
+ * local deletes, we reset the version to 0.
+ *
+ * Here's the expected transition table.
+ *
+ * MUTATION           APPLIED TO      RESULTS IN
+ * SetMutation        Document(v3)    Document(v3)
+ * SetMutation        NoDocument(v3)  Document(v0)
+ * SetMutation        null            Document(v0)
+ * PatchMutation      Document(v3)    Document(v3)
+ * PatchMutation      NoDocument(v3)  NoDocument(v3)
+ * PatchMutation      null            null
+ * TransformMutation  Document(v3)    Document(v3)
+ * TransformMutation  NoDocument(v3)  NoDocument(v3)
+ * TransformMutation  null            null
+ * DeleteMutation     Document(v3)    NoDocument(v0)
+ * DeleteMutation     NoDocument(v3)  NoDocument(v0)
+ * DeleteMutation     null            NoDocument(v0)
+ *
+ * For acknowledged mutations, we use the updateTime of the WriteResponse as the
+ * resulting version for Set, Patch, and Transform mutations. As deletes have no
+ * explicit update time, we use the commitTime of the WriteResponse for
+ * acknowledged deletes.
+ *
+ * If a mutation is acknowledged by the backend but fails the precondition check
+ * locally, we return an `UnknownDocument` and rely on Watch to send us the
+ * updated version.
+ *
+ * Note that TransformMutations don't create Documents (in the case of being
+ * applied to a NoDocument), even though they would on the backend. This is
+ * because the client always combines the TransformMutation with a SetMutation
+ * or PatchMutation and we only want to apply the transform if the prior
+ * mutation resulted in a Document (always true for a SetMutation, but not
+ * necessarily for an PatchMutation).
+ */
+class Mutation {
+ public:
+  virtual ~Mutation() {
+  }
+
+  const DocumentKey& key() const {
+    return key_;
+  }
+  const Precondition& precondition() const {
+    return precondition_;
+  }
+
+  // TODO(rsgowman): ApplyToRemoteDocument()
+
+  /**
+   * Applies this mutation to the given MaybeDocument for the purposes of
+   * computing the new local view of a document. Both the input and returned
+   * documents can be nullptr.
+   *
+   * @param maybe_doc The document to mutate. The input document can be nullptr
+   *     if the client has no knowledge of the pre-mutation state of the
+   *     document.
+   * @param base_doc The state of the document prior to this mutation batch. The
+   *     input document can be nullptr if the client has no knowledge of the
+   *     pre-mutation state of the document.
+   * @param local_write_time A timestamp indicating the local write time of the
+   *     batch this mutation is a part of.
+   * @return The mutated document. The returned document may be nullptr, but
+   *     only if maybe_doc was nullptr and the mutation would not create a new
+   *     document.
+   */
+  virtual std::unique_ptr<MaybeDocument> ApplyToLocalView(
+      const MaybeDocument* maybe_doc,
+      const MaybeDocument* base_doc,
+      const Timestamp& local_write_time) const = 0;
```

Comment:
Very optional: `Timestamp` is just two ints (`int64_t` and `int32_t`) and would probably be more efficient to pass by value.